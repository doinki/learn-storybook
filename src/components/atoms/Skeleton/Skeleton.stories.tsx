import type { ComponentMeta, ComponentStory } from '@storybook/react';

import Skeleton from '.';

export default {
  component: Skeleton,
  title: 'Skeleton',
} as ComponentMeta<typeof Skeleton>;

const Template: ComponentStory<typeof Skeleton> = (args) => (
  <Skeleton {...args} />
);

export const Default = Template.bind({});
Default.args = {
  className: 'w-[240px] h-[24px]',
  variant: 'text',
};

export const Rectangle = Template.bind({});
Rectangle.args = {
  className: 'w-[120px] h-[120px]',
  variant: 'rectangle',
};

export const Circle = Template.bind({});
Circle.args = {
  className: 'w-[120px] h-[120px]',
  variant: 'circular',
};
