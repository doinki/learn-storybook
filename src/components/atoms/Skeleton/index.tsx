import clsx from 'clsx';
import type { FC } from 'react';

interface SkeletonProps {
  className?: string;
  /** @default 'text' */
  variant?: 'text' | 'circular' | 'rectangle';
}

// TODO: animation
const Skeleton: FC<SkeletonProps> = ({ className, variant = 'text' }) => {
  let rounded;

  switch (variant) {
    case 'text':
      rounded = 'rounded-md p-1 bg-clip-content';
      break;
    case 'circular':
      rounded = 'rounded-full';
      break;
    case 'rectangle':
    default:
  }

  return (
    <span
      className={clsx('block bg-gray-200 dark:bg-gray-500', rounded, className)}
      data-testid="skeleton"
    />
  );
};

export default Skeleton;
