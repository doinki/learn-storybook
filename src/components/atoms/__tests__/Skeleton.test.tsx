import { composeStories } from '@storybook/testing-react';
import { render, screen } from '@testing-library/react';

import Skeleton from '../Skeleton';
import * as stories from '../Skeleton/Skeleton.stories';

const { Circle, Default, Rectangle } = composeStories(stories);

describe('atoms/Skeleton', () => {
  test('Have its own classname', () => {
    const className = 'h-64 w-64 bg-black';
    render(<Skeleton className={className} />);
    expect(screen.getByText('', { selector: 'span' })).toHaveClass(className);
  });
  test('No children', () => {
    render(<Skeleton />);
    expect(screen.getByText('', { selector: 'span' })).toBeEmptyDOMElement();
  });
  test('Support dark mode', () => {
    render(<Skeleton />);
    expect(screen.getByText('', { selector: 'span' })).toHaveClass(
      'dark:bg-gray-500'
    );
  });
  test('Default has color only content area', () => {
    render(<Default />);
    expect(screen.getByText('', { selector: 'span' })).toHaveClass(
      'bg-clip-content',
      'p-1',
      'rounded-md'
    );
  });
  test('Circle have rounded-full classname', () => {
    render(<Circle />);
    expect(screen.getByText('', { selector: 'span' })).toHaveClass(
      'rounded-full'
    );
  });

  // snapshots
  test('Default snapshot', () => {
    render(<Default />);
    expect(screen.getByText('', { selector: 'span' })).toMatchSnapshot();
  });
  test('Circle snapshot', () => {
    render(<Circle />);
    expect(screen.getByText('', { selector: 'span' })).toMatchSnapshot();
  });
  test('Rectangle snapshot', () => {
    render(<Rectangle />);
    expect(screen.getByText('', { selector: 'span' })).toMatchSnapshot();
  });
});
