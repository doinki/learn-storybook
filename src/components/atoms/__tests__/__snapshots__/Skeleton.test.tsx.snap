// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Skeleton Circle snapshot 1`] = `
<span
  class="block bg-gray-200 dark:bg-gray-500 rounded-full w-[120px] h-[120px]"
/>
`;

exports[`atoms/Skeleton Default snapshot 1`] = `
<span
  class="block bg-gray-200 dark:bg-gray-500 p-1 rounded-md bg-clip-content w-[240px] h-[24px]"
/>
`;

exports[`atoms/Skeleton Rectangle snapshot 1`] = `
<span
  class="block bg-gray-200 dark:bg-gray-500 w-[120px] h-[120px]"
/>
`;
